server/src/utils.ts

```diff
@@ -143,3 +143,69 @@ export class MruTracker {
 export function tsDisplayPartsToText(parts: ts.SymbolDisplayPart[]): string {
   return parts.map(dp => dp.text).join('');
 }
+
+interface DocumentPosition {
+  fileName: string;
+  pos: number;
+}
+
+/**
+ *
+ * This function attempts to use *internal* TypeScript APIs to find the original source spans for
+ * the `ts.DefinitionInfo` using source maps. If it fails, this function returns the same
+ * `ts.DefinitionInfo` that was passed in.
+ *
+ * @see https://github.com/angular/vscode-ng-language-service/issues/1588
+ */
+export function getMappedDefinitionInfo(
+    info: ts.DefinitionInfo, project: ts.server.Project): ts.DefinitionInfo {
+  try {
+    const mappedDocumentSpan = getMappedDocumentSpan(info, project);
+    return {...info, ...mappedDocumentSpan};
+  } catch {
+    return info;
+  }
+}
+
+function getMappedDocumentSpan(
+    documentSpan: ts.DocumentSpan, project: ts.server.Project): ts.DocumentSpan|undefined {
+  const newPosition = getMappedLocation(documentSpanLocation(documentSpan), project);
+  if (!newPosition) return undefined;
+  return {
+    fileName: newPosition.fileName,
+    textSpan: {start: newPosition.pos, length: documentSpan.textSpan.length},
+    originalFileName: documentSpan.fileName,
+    originalTextSpan: documentSpan.textSpan,
+    contextSpan: getMappedContextSpan(documentSpan, project),
+    originalContextSpan: documentSpan.contextSpan
+  };
+}
```

Comment:
I think we can make this more simple because only the `fileName` and `textSpan` are used in the method `tsDefinitionsToLspLocationLinks`

Reply:
That is true. The `originalTextSpan` is documented with

> "If the span represents a location that was remapped (e.g. via a .d.ts.map file), then the original filename and span will be specified here". 


I don't think there's any harm in keeping these here. Do you have any strong preference either way? I feel think they should be included for correctness/completeness.

Reply:
No, there is no harm in this. It's ok for me. I just tend to add the var we need.

https://github.com/microsoft/TypeScript/blob/818962087cadaae70a44303678ef166c61ca0bb1/src/server/session.ts#L1315-L1336

From the comment, looks like it's used to be compatible with Visual Studio. I'm not familiar with Visual Studio.

Reply:
@ivanwonder Sounds good. Would you mind "officially" reviewing and approving this PR? At this point, I think you're the right person to review most language service PRs.